### PR TITLE
fix string slice & splice

### PR DESCRIPTION
**Pull request by Parashurama**
fixes https://github.com/nim-lang/Nim/issues/4636
primary code fix courtesy of @memophen
added test file for edge cases.

### CODE REVIEW

File: tests/stdlib/tstring.nim [hunk 1]
```diff
@@ -0,0 +1,52 @@
+discard """
+  file: "tstring.nim"
+  output: '''
```
Comment (Araq) [inline tests/stdlib/tstring.nim]: I will accept this, but in general I don't like ``unittest`` for testament. Unnecessary and slows down compilations. There is nothing here that wouldn't work with ``doAssert`` instead.

Comment (Parashurama) [inline tests/stdlib/tstring.nim]: Will change this. I just find it convenient when you need to reset state for each test.
